lib/App/Pinto/Command/pull.pm

```diff
@@ -203,6 +204,51 @@ to work on those distributions, in the event that you need to patch them
 in the future.  Be aware that most distributions do not actually declare
 their development prerequisites.
 
+=item --verify-upstream=[0-5]
+
+=item -Z [0-5]
+
+=item -Z ... -ZZZZZ
+
+!! THIS OPTION IS EXPERIMENTAL !!
+
+Require upstream distribution files to be verified before operating on them.
+Repeated use of this argument (up to 5) requires the upstream verification to
+be progressively more strict.  You can also set the verification level
+explicitly, e.g.,
+
+    --verify-upstream=3
+
+At level 0, no verification is performed. This may be useful if you need to
+override the verification level set earlier, say, in a script.
+
+At level 1, we verify the distributions checksum using the upstream CHECKSUMS
+file.
+
+At level 2, we also verify the signature on the upstream CHECKSUMS file if it
+has one.  Warnings about unknown or untrusted PGP keys relating to that file
+are printed.
+
+At level 3, we also require upstream CHECKSUMS files to be signed.  Warnings
+about unknown or untrusted PGP keys relating to that file are now considered
+fatal.
+
+At level 4, we also verify the unpacked distribution using the embedded
```

Comment:
On 6 February 2015 at 16:23, Jeffrey Ryan Thalhammer <
notifications@github.com> wrote:

> Does that mean if we pull --verify-upstream=4 that the distro will be
> unpacked twice (once to parse for packages, and once to check the
> signatures)?


​There is a subtle trap here.​  Unpacking a corrupted archive is
potentially dangerous, so if we fail the level 3 checks, we want to bail
out before unpacking and running the level 4 checks.  Removing the
duplication would add a fair bit of complexity to the implementation.  Not
impossible, and I think I've just figured out how we could do this, but I
wonder if the small speed up would justify the added complexity, given that
this only kicks in if you have your paranoia level cranked up to 4 or more.